flocker/control/test/test_httpapi.py

```diff
@@ -456,3 +460,259 @@ def test_listens_endpoint(self):
+        expected = []
+        self.assertEqual(expected, list(datasets_from_deployment(deployment)))
+
+    def test_application_volumes(self):
+        """
+        ``datasets_from_deployment`` returns dataset dictionaries for the
+        volumes attached to applications on all nodes.
+        """
+        expected_hostname = u"node1.example.com"
+        expected_dataset = Dataset(dataset_id=u"jalkjlk")
+        volume = AttachedVolume(
+            manifestation=Manifestation(dataset=expected_dataset,
+                                        primary=True),
+            mountpoint=FilePath(b"/blah"))
+
+        node = Node(
+            hostname=expected_hostname,
+            applications=frozenset({Application(name=u'mysql-clusterhq',
+                                                image=object()),
+                                    Application(name=u'site-clusterhq.com',
+                                                image=object(),
+                                                volume=volume)}),
+        )
+
+        deployment = Deployment(nodes=frozenset([node]))
+        expected = dict(
+            dataset_id=expected_dataset.dataset_id,
+            primary=expected_hostname,
+            metadata=thaw(expected_dataset.metadata)
+        )
+        self.assertEqual(
+            [expected], list(datasets_from_deployment(deployment)))
+
+    def test_other_manifestations(self):
+        """
+        ``datasets_from_deployment`` returns dataset dictionaries for the
+        other_manifestations on all nodes.
+        """
+        expected_hostname = u"node1.example.com"
+        expected_dataset = Dataset(dataset_id=u"jalkjlk")
+        expected_manifestation = Manifestation(dataset=expected_dataset,
+                                               primary=True)
+        node = Node(
+            hostname=expected_hostname,
+            applications=frozenset(),
+            other_manifestations=frozenset([expected_manifestation])
+        )
+
+        deployment = Deployment(nodes=frozenset([node]))
+        expected = dict(
+            dataset_id=expected_dataset.dataset_id,
+            primary=expected_hostname,
+            metadata=thaw(expected_dataset.metadata)
+        )
+        self.assertEqual(
+            [expected], list(datasets_from_deployment(deployment)))
+
+    def test_primary_and_secondary_manifestations(self):
+        """
+        ``datasets_from_deployment`` does not return secondary manifestations
+        on other nodes.
+        """
+
+        expected_hostname = u"node1.example.com"
+        expected_dataset = Dataset(dataset_id=u"jalkjlk")
+        volume = AttachedVolume(
+            manifestation=Manifestation(dataset=expected_dataset,
+                                        primary=True),
+            mountpoint=FilePath(b"/blah"))
+
+        node1 = Node(
+            hostname=expected_hostname,
+            applications=frozenset({Application(name=u'mysql-clusterhq',
```

Comment:
No more mysql examples in the test suite, please. :smile:  Sorry about the bad examples that inspired this.